File: problem-set-2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
int calculate_index(int letters, int words, int sentences);
void print_grade(int index);

int main(void)
{
    // Prompt the user for some text
    string text = get_string("Text: ");

    // Count the number of letters, words, and sentences in the text
    int letter_count = count_letters(text);
    int word_count = count_words(text);
    int sentence_count = count_sentences(text);

    // Compute the Coleman-Liau index
    int index = calculate_index(letter_count, word_count, sentence_count);

    // Print the grade level
    print_grade(index);
}

int count_letters(string text)
{
    int letter_count = 0;

    for (int i = 0, length = strlen(text); i < length; i++)
    {
        if (isalpha(text[i]))
        {
            letter_count += 1;
        }
    }

    return letter_count;
}

int count_words(string text)
{
    int word_count = 1; // At least one word in a non-empty text

    for (int i = 0, length = strlen(text); i < length; i++)
    {
        if (isspace(text[i]))
        {
            word_count += 1;
        }
    }

    return word_count;
}

int count_sentences(string text)
{
    int sentence_count = 0;

    for (int i = 0, length = strlen(text); i < length; i++)
    {
        char currentChar = text[i];

        if (currentChar == '.' || currentChar == '!' || currentChar == '?')
        {
            sentence_count += 1;
        }
    }

    return sentence_count;
}

int calculate_index(int letters, int words, int sentences)
{
    float L = (float) letters / words * 100;
    float S = (float) sentences / words * 100;

    return round(0.0588 * L - 0.296 * S - 15.8);
}

void print_grade(int index)
{
    if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

File: problem-set-6/readability/readability.py
def main():
    # Prompt the user for some text
    text = input("Text: ")

    # Count the number of letters, words, and sentences in the text
    letter_count = count_letters(text)
    word_count = count_words(text)
    sentence_count = count_sentences(text)

    # Compute the Coleman-Liau index
    index = calculate_index(letter_count, word_count, sentence_count)

    # Print the grade level
    print_grade(index)


def count_letters(text):
    letter_count = 0

    for character in text:
        if character.isalpha():
            letter_count += 1

    return letter_count


def count_words(text):
    word_count = 1  # At least one word in a non-empty text

    for character in text:
        if character.isspace():
            word_count += 1

    return word_count


def count_sentences(text):
    sentence_count = 0

    for character in text:
        if character == "." or character == "!" or character == "?":
            sentence_count += 1

    return sentence_count


def calculate_index(letters, words, sentences):
    l = letters / words * 100
    s = sentences / words * 100

    return round(0.0588 * l - 0.296 * s - 15.8)


def print_grade(index):
    if index >= 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {index}")


main()
